cram_pr2/cram_pr2_projection/src/tf.lisp

```diff
@@ -43,43 +44,81 @@
-     (cl-tf:transform->transform-stamped
-      odom-frame base-frame time
-      (cl-transforms:pose->transform robot-pose-in-map))
-     :suppress-callbacks t)
+      ;; tell the tf transformer where the robot is in the odometry frame
+      (cl-tf:set-transform
+       transformer
+       (cl-tf:transform->transform-stamped
+        odom-frame base-frame time
+        (cl-transforms:pose->transform robot-pose-in-map))
+       :suppress-callbacks t)
 
-    ;; tell the tf transformer the current configuration of robot's joints
-    (dolist (link (btr:link-names ?robot-instance))
-      (unless (equal link base-frame)
-        (let ((transform (cl-transforms:transform*
-                          reference-transform-inv
-                          (cl-transforms:reference-transform
-                           (btr:link-pose ?robot-instance link)))))
-          (cl-tf:set-transform
-           transformer
-           (cl-tf:make-transform-stamped
-            base-frame link time
-            (cl-transforms:translation transform)
-            (cl-transforms:rotation transform))
-           :suppress-callbacks t))))
+      ;; tell the tf transformer the current configuration of robot's joints
+      (dolist (link (btr:link-names ?robot-instance))
+        (unless (equal link base-frame)
+          (let ((transform (cl-transforms:transform*
+                            reference-transform-inv
+                            (cl-transforms:reference-transform
+                             (btr:link-pose ?robot-instance link)))))
+            (cl-tf:set-transform
+             transformer
+             (cl-tf:make-transform-stamped
+              base-frame link time
+              (cl-transforms:translation transform)
+              (cl-transforms:rotation transform))
+             :suppress-callbacks t))))
 
-    ;; execute the TF update callbacks
-    (cl-tf:execute-changed-callbacks transformer))))
+      ;; tell the tf transformer about the virtual fixed joints
+      (let* ((bullet-links (btr:link-names ?robot-instance))
+             (urdf-links (cl-urdf:links (btr:urdf ?robot-instance)))
+             ;; get the names of the links, which are only present in the urdf
+             (urdf-link-names (remove-if (lambda (x) (member x bullet-links))
+                                         (loop for name being the hash-keys in urdf-links
+                                               collecting name))))
+        (let ((prev-link-count))
+          ;; while there are still links left to be asserted in tf
+          (loop while (> (length urdf-link-names) 0)
+                do ;; if no links were asserted last loop, we are stuck and have to return
+                   (if (eq prev-link-count (length urdf-link-names))
+                       (progn
+                           (roslisp:ros-error (pr2-proj tf) "Somehow a virtual link with no parents in the bullet world was found in the urdf. This should never happen.")
+                           (return))
```

Comment:
Does the `return` exit from the whole function? Is that correct behavior? Should it not simply exit the virtual link assertion part?

Reply:
As I tested it `return` exits the loop macro. The HyperSpec states it returns from the block named `nil`, which is the name of implicitly created blocks like from `do`. I can't test this right now, so this would have to wait until next week. But I think it works as I intended it to.